### DISCUSSION TITLE
Fragment SMILES are not canonical

### DISCUSSION THREAD

**Discussion by laurenreid1** [rdkit/rdkit#7727] [Q&A]
Hi RDKit community,

RDKit version: 2023.9.5 

I have a process that requires canonicalisation of fragment SMILES. I have noticed the following behaviour when trying to canonicalise the following fragment:

```
>>> mol_a = Chem.MolFromSmiles("cCC", sanitize=False)
>>> mol_b = Chem.MolFromSmiles("CCc", sanitize=False)
>>> print(Chem.MolToSmiles(mol_a, canonical=True))
cCC
>>> print(Chem.MolToSmiles(mol_b, canonical=True))
CCc
```

Could this be a bug or is it because it's not possible to canonicalise SMILES of fragments that are unable to be sanitized? 

Thanks,

Lauren

**Comment by laurenreid1** [comment]
Hi RDKit community,

I'm just bumping this issue. I have also noticed the behaviour with the following fragments:

```
>>> mol = Chem.MolFromSmiles("CC(c)CN", sanitize=False)
>>> Chem.MolToSmiles(mol)
'CC(c)CN'
>>> mol = Chem.MolFromSmiles("cC(C)CN", sanitize=False)
>>> Chem.MolToSmiles(mol)
'cC(C)CN'
```

I appreciate the fragments are not sanitized when they're parsed in (and are obviously not chemically feasible) but my process involves extracting fragments like these and storing them in a canonicalised database

**Comment by ptosco** [comment]
@laurenreid1 Hi Lauren, the problem is that the aromatic carbon in the fragment stays `SP3`-hybridized even if partial sanitization is carried out omitting only the `SANITIZE_KEKULIZE` step.
Therefore, both `c` aromatic carbons are assigned 3 implicit Hs as if they were methyl carbons. The calculation of atom canonical ranks uses the number of implicit Hs to disambiguate, and does not seem to consider the `GetIsAromatic()` flag at all:
```
mol_a = Chem.MolFromSmiles("cCC", sanitize=False)
Chem.SanitizeMol(mol_a, Chem.SanitizeFlags.SANITIZE_ALL ^ Chem.SanitizeFlags.SANITIZE_KEKULIZE)
for a in mol_a.GetAtoms():
    print(f"{a.GetIdx()} {a.GetSymbol()} arom {a.GetIsAromatic()} nHs {a.GetTotalNumHs()} hyb {a.GetHybridization()}")
print(list(Chem.CanonicalRankAtoms(mol_a, breakTies=False)), Chem.MolToSmiles(mol_a))
0 C arom True nHs 3 hyb SP3
1 C arom False nHs 2 hyb SP3
2 C arom False nHs 3 hyb SP3
[0, 2, 0] cCC

mol_b = Chem.MolFromSmiles("CCc", sanitize=False)
Chem.SanitizeMol(mol_b, Chem.SanitizeFlags.SANITIZE_ALL ^ Chem.SanitizeFlags.SANITIZE_KEKULIZE)
for a in mol_b.GetAtoms():
    print(f"{a.GetIdx()} {a.GetSymbol()} arom {a.GetIsAromatic()} nHs {a.GetTotalNumHs()} hyb {a.GetHybridization()}")
print(list(Chem.CanonicalRankAtoms(mol_b, breakTies=False)), Chem.MolToSmiles(mol_b))
0 C arom False nHs 3 hyb SP3
1 C arom False nHs 2 hyb SP3
2 C arom True nHs 3 hyb SP3
[0, 2, 0] CCc
```
Given the two atoms have the same canonical rank, unsurprisingly they cannot be uniquely canonicalized.
I would incline to consider this a bug - an aromatic carbon should never be `SP3`-hybridized, but @greglandrum might have a different opinion.

A possible workaround is to manually set hybridization of aromatic atoms in the fragment to `SP2` and add a radical electron. By doing so, only two implicit Hs are assigned to the aromatic carbon, and therefore the same canonical SMILES is then generated:
```
def massage_fragment(mol):
    Chem.SanitizeMol(mol, Chem.SanitizeFlags.SANITIZE_ALL ^ Chem.SanitizeFlags.SANITIZE_KEKULIZE)
    for a in mol.GetAtoms():
        if a.GetIsAromatic():
            a.SetHybridization(Chem.HybridizationType.SP2)
            a.SetNumRadicalElectrons(1)
    mol.UpdatePropertyCache(True)

mol_a = Chem.MolFromSmiles("cCC", sanitize=False)
massage_fragment(mol_a)
for a in mol_a.GetAtoms():
    print(f"{a.GetIdx()} {a.GetSymbol()} arom {a.GetIsAromatic()} nHs {a.GetTotalNumHs()} hyb {a.GetHybridization()}")
print(list(Chem.CanonicalRankAtoms(mol_a, breakTies=False)), Chem.MolToSmiles(mol_a))
0 C arom True nHs 2 hyb SP2
1 C arom False nHs 2 hyb SP3
2 C arom False nHs 3 hyb SP3
[0, 2, 1] [cH2]CC

mol_b = Chem.MolFromSmiles("CCc", sanitize=False)
massage_fragment(mol_b)
for a in mol_b.GetAtoms():
    print(f"{a.GetIdx()} {a.GetSymbol()} arom {a.GetIsAromatic()} nHs {a.GetTotalNumHs()} hyb {a.GetHybridization()}")
print(list(Chem.CanonicalRankAtoms(mol_b, breakTies=False)), Chem.MolToSmiles(mol_b))
0 C arom False nHs 3 hyb SP3
1 C arom False nHs 2 hyb SP3
2 C arom True nHs 2 hyb SP2
[1, 2, 0] [cH2]CC
```

**Comment by laurenreid1** [comment]
Hi @ptosco,

Aha, that makes sense. Thanks for the workaround, I'll give that a go!

Lauren
